vcr/issuer/networkPublisher.go

```diff
@@ -0,0 +1,132 @@
+// It is the default implementation to use for issuers to publish credentials and revocations to the Nuts network.
+func NewNetworkPublisher(networkTx network.Transactions, docResolver vdr.DocResolver, keyResolver crypto.KeyResolver) Publisher {
+	return &networkPublisher{
+		networkTx:       networkTx,
+		didDocResolver:  docResolver,
+		serviceResolver: doc.NewServiceResolver(docResolver),
+		keyResolver: vdrKeyResolver{
+			docResolver: docResolver,
+			keyResolver: keyResolver,
+		},
+	}
+
+}
+
+func (p networkPublisher) PublishCredential(verifiableCredential vc.VerifiableCredential, public bool) error {
+	issuerDID, err := did.ParseDIDURL(verifiableCredential.Issuer.String())
+	if err != nil {
+		return fmt.Errorf("invalid credential issuer: %w", err)
+	}
+
+	if len(verifiableCredential.CredentialSubject) == 0 {
+		return fmt.Errorf("missing credentialSubject")
+	}
+
+	// create participants list
+	participants, err := p.generateParticipants(verifiableCredential, public)
+	if err != nil {
+		return err
+	}
+
+	key, err := p.keyResolver.ResolveAssertionKey(*issuerDID)
+	if err != nil {
+		return fmt.Errorf("could not resolve an assertion key for issuer: %w", err)
+	}
+
+	// find did document/metadata for originating TXs
+	_, meta, err := p.didDocResolver.Resolve(*issuerDID, nil)
+	if err != nil {
+		return err
+	}
+
+	payload, _ := json.Marshal(verifiableCredential)
+	tx := network.TransactionTemplate(VcDocumentType, payload, key).
+		WithTimestamp(verifiableCredential.IssuanceDate).
+		WithAdditionalPrevs(meta.SourceTransactions).
+		WithPrivate(participants)
+
+	_, err = p.networkTx.CreateTransaction(tx)
+	if err != nil {
+		return fmt.Errorf("failed to publish credential, error while creating transaction: %w", err)
+	}
+	log.Logger().Infof("Verifiable Credential published (id=%s,type=%s)", verifiableCredential.ID, verifiableCredential.Type)
+
+	return nil
+}
+
+func (p networkPublisher) generateParticipants(verifiableCredential vc.VerifiableCredential, public bool) ([]did.DID, error) {
+	issuer, _ := did.ParseDID(verifiableCredential.Issuer.String())
+	participants := make([]did.DID, 0)
+	if !public {
+		var (
+			base                []credential.BaseCredentialSubject
+			credentialSubjectID *did.DID
+		)
+		err := verifiableCredential.UnmarshalCredentialSubject(&base)
+		if err == nil {
+			credentialSubjectID, err = did.ParseDID(base[0].ID) // earlier validation made sure length == 1 and ID is present
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine credentialSubject.ID: %w", err)
+		}
+
+		// participants are not the issuer and the credentialSubject.id but the DID that holds the concrete endpoint for the NutsComm service
+		for _, vcp := range []did.DID{*issuer, *credentialSubjectID} {
+			serviceOwner, err := p.resolveNutsCommServiceOwner(vcp)
+			if err != nil {
+				return nil, fmt.Errorf("could not publish private credential: failed to resolve participating node (did=%s): %w", vcp.String(), err)
```

Comment:
The `could not publish` part should be added by the caller of this method. This method can't possibly know it's used for a single purpose.